R/mobr.R

```diff
@@ -10,221 +10,244 @@ require(pracma)
+        sp_bool = as.data.frame((sp_ordered == 0) * 1) 
         rich = cumprod(sp_bool)
-        explic_loop[ , i] = as.numeric(ncol(dat_sp) - 1 - rowSums(rich))
+        explicit_loop[ , i] = as.numeric(ncol(dat_sp) - 1 - rowSums(rich))
     }
-    explic_S = apply(explic_loop, 1, mean)
-    return(explic_S)
+    explicit_S = apply(explicit_loop, 1, mean)
+    return(explicit_S[effort])
 }
 
-rarefy_sample_implicit = function(dat_sp, dat_plot, treatment) {
-    plot_trmt = dat_plot[dat_plot[ , 2] == treatment, ]
-    sp_trmt = dat_sp[match(plot_trmt[ , 1], dat_sp[ , 1]), ]
-    sample_S = rarefaction.sample(sp_trmt[ , 2:ncol(sp_trmt)])[ , 2]
+rarefy_sample_implicit = function(dat_sp, dat_plot, group, effort=NULL) {
+    sp_grp = dat_sp[dat_plot$group == group, ]
+    sample_S = rarefaction(sp_grp, 'samp', effort)
     return(sample_S)
 }
 
-rarefy_individual = function(dat_sp, dat_plot, treatment) {
-    # Returns the rarefaction results at each sample size
-    sp_trmt = dat_sp[match(dat_plot[dat_plot[ , 2] == treatment, 1], dat_sp[ , 1]), ]
-    trmt_sad = as.numeric(apply(sp_trmt[ , 2:ncol(sp_trmt)], 2, sum))
-    trmt_sad = trmt_sad[trmt_sad != 0]
-    ind_S = rarefaction.individual(trmt_sad)[ , 2]
+rarefy_individual = function(dat_sp, dat_plot, group, effort=NULL) {
+    sad = colSums(dat_sp[dat_plot$group == group, ])
+    ind_S = rarefaction(sad, 'indiv', effort)
     return(ind_S)
 }
 
-get_deltaSsad = function(dat_sp, dat_plot, treatment1, treatment2) {
-    rescaled_ind = sapply(c(treatment1, treatment2), function(x) 
-                          rarefy_individual(dat_sp, dat_plot, x))
-    Nind = min(length(unlist(rescaled_ind[1])), length(unlist(rescaled_ind[2])))
-    deltaSsad = as.numeric(unlist(rescaled_ind[2])[1:Nind] - 
-                           unlist(rescaled_ind[1])[1:Nind])
+get_deltaSsad = function(dat_sp, dat_plot, groups, Nind=NULL) {
+    if (is.null(Nind))
+        Nind = min(tapply(rowSums(dat_sp),  dat_plot$group, sum))
+    # the first element of the argument groups must be the reference group
+    rescaled_ind = sapply(groups, function(x) 
+                          rarefy_individual(dat_sp, dat_plot, x, 1:Nind))
+    deltaSsad = rescaled_ind[ , 2] - rescaled_ind[ , 1]
     return(deltaSsad)
 }
 
-get_deltaSN = function(dat_sp, dat_plot, treatment1, treatment2, ScaleBy = NA) {
-    avg_dens = get_avg_dens(dat_sp, dat_plot, ScaleBy)
-    nplots = c(nrow(dat_plot[dat_plot[ , 2] == treatment1, ]), nrow(dat_plot[dat_plot[ , 
-        2] == treatment2, ]))
-    implicit_sample = sapply(c(treatment1, treatment2), function(x)
-                             rarefy_sample_implicit(dat_sp, dat_plot, x)[1:min(nplots)])
+get_deltaSN = function(dat_sp, dat_plot, groups, ScaleBy = NA, Nind=NULL) {
+    if (is.null(Nind))
+        Nind = min(tapply(rowSums(dat_sp),  dat_plot$group, sum))
+    avg_dens = get_avg_dens(dat_sp, dat_plot, ScaleBy) 
+    # shouldn't the above line be:  'mean(rowSums(dat_sp))' ?
```

Comment:
I think it is pretty important to know which group is specified as the reference otherwise your interpretation of all the graphs could be potentially backward. So either we should not allow the user to not specify `ScaleBy` or we should indicate clearly in outputting tables and figures which group was the reference group. I'd lean towards the first option which is much similar to implement.